### PR TITLE
Added e2e testing framework

### PR DESCRIPTION
**Pull request by yuchaoran2011**
Addressed #307.

I did a `git merge --squash` to clean up the history. Let's see if the CI passes this time.

### CODE REVIEW

**Comment by yuchaoran2011** [comment]
@liyinan926 Now I have no idea why `go fmt` is complaining. I created the branch for this PR directly off of the latest master. Formatting check didn't report any issues on my local machine.

**Comment by yuchaoran2011** [comment]
@liyinan926 CI just passed. Go version was the issue. I created #310 for that.

File: Gopkg.toml [hunk 1]
```diff
@@ -49,6 +49,22 @@ required = [
   name = "github.com/google/go-cloud"
   version = "0.1.0"
 
+[[override]]
```
Comment (liyinan926) [inline Gopkg.toml]: These already exist down below in the file.

File: test/e2e/framework/cluster_role_binding.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+)
+
+func CreateClusterRoleBinding(kubeClient kubernetes.Interface, ns string, relativePath string) (finalizerFn, error) {
+	finalizerFn := func() error { return DeleteClusterRoleBinding(kubeClient, relativePath) }
```
Comment (liyinan926) [inline test/e2e/framework/cluster_role_binding.go]: Please avoid this kind of one line function for consistency reason.

File: test/e2e/main_test.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	defer func() {
+		if err := framework.Teardown(); err != nil {
+			log.Fatalf("failed to tear down framework: %v\n", err)
+		}
+		os.Exit(code)
+	}()
+
+	code = m.Run()
```
Comment (liyinan926) [inline test/e2e/main_test.go]: Should it be `os.Exit(m.Run())`?

File: test/e2e/main_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+	if framework, err = operatorFramework.New(*ns, *kubeconfig, *opImage); err != nil {
+		log.Fatalf("failed to set up framework: %v\n", err)
+	}
+
+	defer func() {
+		if err := framework.Teardown(); err != nil {
+			log.Fatalf("failed to tear down framework: %v\n", err)
+		}
+		os.Exit(code)
```
Comment (liyinan926) [inline test/e2e/main_test.go]: It doesn't look like `code` is defined before this line. Do we really need that `code` var?

File: test/e2e/main_test.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+var framework *operatorFramework.Framework
+
+func TestMain(m *testing.M) {
+	kubeconfig := flag.String("kubeconfig", "", "kube config path, e.g. $HOME/.kube/config")
+	opImage := flag.String("operator-image", "", "operator image, e.g. image:tag")
+	ns := flag.String("namespace", "spark-operator", "e2e test namespace")
+	flag.Parse()
+
+	var (
+		err  error
```
Comment (liyinan926) [inline test/e2e/main_test.go]: Looks like we don't really need those two vars.

File: test/e2e/framework/role_binding.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+)
+
+func CreateRoleBinding(kubeClient kubernetes.Interface, ns string, relativePath string) (finalizerFn, error) {
+	finalizerFn := func() error { return DeleteRoleBinding(kubeClient, ns, relativePath) }
```
Comment (liyinan926) [inline test/e2e/framework/role_binding.go]: Also this and other places.

File: test/e2e/main_test.go (hunk 5 above)
Comment (yuchaoran2011) [reply test/e2e/main_test.go]: The `err` was needed in this line:
`if framework, err = operatorFramework.New(*ns, *kubeconfig, *opImage); err != nil`
The `code` var is now removed.

File: test/e2e/main_test.go [hunk 7]
```diff
@@ -33,6 +33,10 @@ func TestMain(m *testing.M) {
 	ns := flag.String("namespace", "spark-operator", "e2e test namespace")
 	flag.Parse()
 
+	var (
+		err error
```
Comment (liyinan926) [inline test/e2e/main_test.go]: You don't really need this explicit var.

Comment (liyinan926) [reply test/e2e/main_test.go]: You can define `err` when you first use it below.

**Pull request merged**